util/stmtsummary/statement_summary.go

```diff
@@ -219,12 +231,64 @@ func (ssMap *stmtSummaryByDigestMap) ToDatum() [][]types.Datum {
 	return rows
 }
 
-// OnEnableStmtSummaryModified is triggered once EnableStmtSummary is modified.
-func OnEnableStmtSummaryModified(newValue string) {
-	if variable.TiDBOptOn(newValue) {
-		atomic.StoreInt32(&variable.EnableStmtSummary, 1)
+// SetDefaultEnabled sets the enabling status by default.
+func (ssMap *stmtSummaryByDigestMap) SetDefaultEnabled(defaultEnabled bool) {
+	ssMap.enabledWrapper.Lock()
+	ssMap.enabledWrapper.defaultEnabled = defaultEnabled
+	ssMap.enabledWrapper.Unlock()
+}
+
+// SetEnabled enables or disables statement summary in global(cluster) or session(server) scope.
+func (ssMap *stmtSummaryByDigestMap) SetEnabled(enable string, inSession bool) {
+	ssMap.enabledWrapper.Lock()
+	needClear := false
+	if inSession {
+		ssMap.enabledWrapper.sessionEnabled = enable
+		if ssMap.isSet(enable) {
+			needClear = !ssMap.isEnabled(enable)
+		} else {
+			needClear = !ssMap.isEnabled(ssMap.enabledWrapper.globalEnabled)
+		}
 	} else {
-		atomic.StoreInt32(&variable.EnableStmtSummary, 0)
-		StmtSummaryByDigestMap.Clear()
+		ssMap.enabledWrapper.globalEnabled = enable
+		if !ssMap.isSet(ssMap.enabledWrapper.sessionEnabled) {
+			needClear = !ssMap.isEnabled(enable)
+		}
 	}
+	ssMap.enabledWrapper.Unlock()
+
+	// Clear all summaries once statement summary is disabled.
+	if needClear {
+		ssMap.Clear()
+	}
+}
+
+// Enabled returns whether statement summary is enabled.
+func (ssMap *stmtSummaryByDigestMap) Enabled() bool {
+	ssMap.enabledWrapper.RLock()
+	var enabled bool
+	if ssMap.isSet(ssMap.enabledWrapper.sessionEnabled) {
+		enabled = ssMap.isEnabled(ssMap.enabledWrapper.sessionEnabled)
+	} else {
+		enabled = ssMap.isEnabled(ssMap.enabledWrapper.globalEnabled)
+	}
+	ssMap.enabledWrapper.RUnlock()
+	return enabled
+}
+
+// isEnabled converts a string value to bool.
+func (ssMap *stmtSummaryByDigestMap) isEnabled(enable string) bool {
+	switch {
+	case strings.EqualFold(enable, "ON") || enable == "1":
```

Comment:
We can covert "ON" to "1" when setting the value.

Reply:
Seems already converted in `ValidateSetSystemVar`.
We can simply return `enable == "1"`.

Reply:
`ValidateSetSystemVar` is executed in `variable.SetSessionSystemVar`, but `variable.SetSessionSystemVar` doesn't return converted value in `SetExecutor.setSysVariable`,  so `stmtsummary.StmtSummaryByDigestMap` gets a raw value and a conversion is needed.

Reply:
I added `normalizeEnableValue` to convert.